### PR TITLE
[SPARK-46995][SQL] Allow AQE coalesce final stage in SQL cached plan

### PR DESCRIPTION
**Pull request by liuzqt**
### What changes were proposed in this pull request?

https://github.com/apache/spark/pull/43435 and https://github.com/apache/spark/pull/43760 are fixing a correctness issue which will be triggered when AQE applied on cached query plan, specifically, when AQE coalescing the final result stage of the cached plan.

 

The current semantic of `spark.sql.optimizer.canChangeCachedPlanOutputPartitioning`

([source code](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala#L403-L411)):

when true, we enable AQE, but disable coalescing final stage (default)
when false, we disable AQE
 

But let’s revisit the semantic of this config: actually for caller the only thing that matters is whether we change the output partitioning of the cached plan. And we should only try to apply AQE if possible.  Thus we want to modify the semantic of spark.sql.optimizer.canChangeCachedPlanOutputPartitioning

when true, we enable AQE and allow coalescing final: this might lead to perf regression, because it introduce extra shuffle
when false, we enable AQE, but disable coalescing final stage. (this is actually the `true` semantic of old behavior)
Also, to keep the default behavior unchanged, we might want to flip the default value of spark.sql.optimizer.canChangeCachedPlanOutputPartitioning to `false`

### Why are the changes needed?

To allow AQE coalesce final stage in SQL cached plan. Also make the semantic of `spark.sql.optimizer.canChangeCachedPlanOutputPartitioning` more reasonable.

### How was this patch tested?

Updated UTs.

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by liuzqt** [comment]
@cloud-fan @maryannxue Please help review this change, thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -1561,11 +1561,11 @@ object SQLConf {
       .doc("Whether to forcibly enable some optimization rules that can change the output " +
         "partitioning of a cached query when executing it for caching. If it is set to true, " +
         "queries may need an extra shuffle to read the cached data. This configuration is " +
-        "enabled by default. The optimization rules enabled by this configuration " +
-        s"are ${ADAPTIVE_EXECUTION_ENABLED.key} and ${AUTO_BUCKETED_SCAN_ENABLED.key}.")
+        "enabled by default. The optimization rule enabled by this configuration " +
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1564]:
```suggestion
        "disabled by default. The optimization rule enabled by this configuration " +
```

File: sql/core/src/test/scala/org/apache/spark/sql/CachedTableSuite.scala [hunk 2]
```diff
@@ -1630,20 +1630,37 @@ class CachedTableSuite extends QueryTest with SQLTestUtils
...
+        } else {
+          root.children.flatMap(findNodeInSparkPlanInfo(_, cond)).headOption
+        }
+      }
+
+      def cachedFinalStageCoalesced(sparkPlanInfo: SparkPlanInfo): Boolean = {
+        val inMemoryScanNode = findNodeInSparkPlanInfo(sparkPlanInfo,
+          _.nodeName.contains("TableCacheQueryStage"))
+        val resultNode = findNodeInSparkPlanInfo(inMemoryScanNode.get,
+          _.nodeName.contains("ResultQueryStage"))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/CachedTableSuite.scala:1660]: there is no `ResultQueryStage` in Spark.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
cc @yaooqinn

**Review by yaooqinn — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by cloud-fan** [comment]
@liuzqt can you correct the JIRA ticket ID? It seems wrong.

**Pull request closed without merging**